wdl-lsp/src/lib.rs

```diff
@@ -0,0 +1,485 @@
+use tower_lsp::LspService;
+use uuid::Uuid;
+use wdl_analysis::Analyzer;
+use wdl_analysis::DocumentChange;
+use wdl_analysis::ProgressKind;
+use wdl_ast::Validator;
+use wdl_lint::LintVisitor;
+use workspace::ClientSupport;
+use workspace::Workspace;
+
+mod proto;
+mod workspace;
+
+/// Constructs a new validator.
+///
+/// This is used as a callback to the analyzer.
+fn validator(lint: bool) -> Validator {
+    let mut validator = Validator::default();
+    if lint {
+        validator.add_visitor(LintVisitor::default());
+    }
+
+    validator
+}
+
+/// Takes the requested changes for the given document.
+///
+/// This is used as a callback to the analyzer.
+fn changes(workspace: &Arc<RwLock<Workspace>>, uri: &Url) -> Option<DocumentChange> {
+    let mut workspace = workspace.write();
+    if let Some(document) = workspace.documents.get_mut(uri) {
+        return document.take_change();
+    }
+
+    None
+}
+
+/// Represents options for running the LSP server.
+#[derive(Debug, Default)]
+pub struct ServerOptions {
+    /// The name of the server.
+    ///
+    /// Defaults to `wdl-lsp` crate name.
+    pub name: Option<String>,
+
+    /// The version of the server.
+    ///
+    /// Defaults to the version of the `wdl-lsp` crate.
+    pub version: Option<String>,
+
+    /// Whether or not linting is enabled.
+    pub lint: bool,
+}
+
+/// Represents an LSP server for analyzing WDL documents.
+#[derive(Debug)]
+pub struct Server {
+    /// The LSP client connected to the server.
+    client: Client,
+    /// The options for the server.
+    options: ServerOptions,
+    /// The analyzer used to analyze documents.
+    analyzer: Analyzer,
+    /// The workspace managed by the server.
+    workspace: Arc<RwLock<Workspace>>,
+}
+
+impl Server {
+    /// Runs the server until a request is received to shut down.
+    pub async fn run(options: ServerOptions) -> Result<()> {
+        log::debug!("running LSP server: {options:#?}");
+
+        let (service, socket) = LspService::new(|client| {
+            let lint = options.lint;
+            let workspace: Arc<RwLock<Workspace>> = Default::default();
+            let analyzer_client = client.clone();
+            let analyzer_workspace = workspace.clone();
+
+            Self {
+                client,
+                options,
+                analyzer: Analyzer::new_with_changes(
+                    move |kind, current, total, context| {
+                        Self::report_analysis_progress(
+                            analyzer_client.clone(),
+                            context.clone(),
+                            kind,
+                            current,
+                            total,
+                        )
+                    },
+                    move || validator(lint),
+                    move |uri| changes(&analyzer_workspace, uri),
+                ),
+
+                workspace,
+            }
+        });
+
+        let stdin = tokio::io::stdin();
```

Comment:
I'm not sure if we want to give some thought as to how to convey the communication method (https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#implementationConsiderations)? Here, we've hardcoded the `stdio` method—I know that's what a lot of implementations do, but do we want to make that more apparent in the initialization somehow?

Regardless, I still recommend `sprocket` eventually require the `--stdio` flag to make it explicit, even if there is only one option.

Reply:
The forthcoming `sprocket analyzer` command implementation does take a `--stdio` option.

I plan on extending the construction of the server to take the transport when I implement tests, so we can spawn the server on as a tokio task and hand it an in-memory communication channel.